CatsApp/Data/Services/DatabaseService.swift

```diff
@@ -0,0 +1,119 @@
+//
+//  BreedsDataService.swift
+//  CatsApp
+//
+//  Created by Carlos Costa on 10/10/2025.
+//
+import Foundation
+import SwiftData
+
+protocol DatabaseServiceProtocol {
+
+    @MainActor
+    func saveBreeds(_ dtos: [BreedsDataService.CatBreed], context: ModelContext)
+        throws -> [CatBreed]
+    @MainActor
+    func fetchAllBreeds(context: ModelContext) throws -> [CatBreed]
+    @MainActor
+    func fetchFavoriteBreeds(context: ModelContext) throws -> [CatBreed]
+    @MainActor
+    func toggleFavorite(_ breed: CatBreed, context: ModelContext) throws
+    @MainActor
+    func cacheImage(forBreedId breedId: String, withUrl urlString: String, context: ModelContext) async throws
+}
+
+struct DatabaseService: DatabaseServiceProtocol {
+
+    @MainActor
+    func saveBreeds(_ dtos: [BreedsDataService.CatBreed], context: ModelContext)
+        throws -> [CatBreed]
+    {
+        guard !dtos.isEmpty else { return try fetchAllBreeds(context: context) }
+        let existingDescriptor = FetchDescriptor<CatBreed>()
+        let existing: [CatBreed]
+        do {
+            existing = try context.fetch(existingDescriptor)
+        } catch {
+            throw DomainError.persistenceError
+        }
+
+        var existingIds = Set(existing.map { $0.id })
+
+        var inserted = false
+        for dto in dtos {
```

Comment:
```suggestion
       dtos.map { dto in 
```

I think you can achieve this with map, which will basically transforms your dtos into an array of CatBreeds for example. I think you still can do all the operations you did inside the body of the for

Reply:
Also, for future knowledge, you also have available `.compactMap` which will return only non-nil elements from your transformation, and now that i'm saying this, maybe is more appropriate for your use-case, since I think `.map` actually returns an optional collection. Or at least, a collection of optional CatBreeds 😅. Let me know if it didn't make any sense.

Reply:
Thanks for the suggestion! 🙌 
I’ll update it to use compactMap instead, it actually fits my case better since I only want to keep the non-nil results and then map them into new CatBreed models!